haystack/document_stores/pinecone.py

```diff
@@ -243,19 +229,128 @@ def _create_index(
+        if not filters:
+            filters = {}
+        return filters
+
+    def _add_type_metadata_filter(self, filters: FilterType, type_value: Optional[str]) -> FilterType:
+        """
+        Add new filter for `doc_type` metadata field.
+        """
+        if type_value:
+            new_type_filter = {self.type_metadata_field: {"$eq": type_value}}
+            if "$and" in filters:
+                if self.type_metadata_field in filters["$and"]:  # type: ignore
+                    current_type_filter = filters["$and"][self.type_metadata_field]  # type: ignore
+                    type_values = [type_value]
+                    if isinstance(current_type_filter, str):
+                        type_values.append(current_type_filter)
+                    elif isinstance(current_type_filter, dict):
+                        if "$eq" in current_type_filter:
+                            # current `doc_type` filter has single value
+                            type_values.append(current_type_filter["$eq"])
+                        else:
+                            # current `doc_type` filter has multiple values
+                            type_values.extend(current_type_filter["$in"])
+                    new_type_filter = {self.type_metadata_field: {"$in": type_values}}  # type: ignore
+                filters["$and"].update(new_type_filter)  # type: ignore
+            else:
+                # extend filters with new `doc_type` filter and add $and operator
+                filters.update(new_type_filter)
+                all_filters = filters
+                filters = {"$and": all_filters}
+        return filters
+
+    def _set_default_type_metadata(self, index: Optional[str], namespace: Optional[str] = None) -> str:
+        """
+        Set default value for `doc_type` metadata filed. If there is at least one embedding, default value
+        will be `vector`, otherwise it will be `no-vector`.
+        """
+        if self.get_embedding_count(index=index, namespace=namespace) > 0:
+            type_metadata_value = self.document_with_embedding_metadata
+        else:
+            type_metadata_value = self.document_without_embedding_metadata
+        return type_metadata_value
+
+    def _get_vector_count(self, index: str, filters: Optional[FilterType], namespace: Optional[str]) -> int:
+        res = self.pinecone_indexes[index].query(
+            self.dummy_query,
+            top_k=self.top_k_limit,
+            include_values=False,
+            include_metadata=False,
+            filter=filters,
+            namespace=namespace,
+        )
+        return len(res["matches"])
+
     def get_document_count(
         self,
         filters: Optional[FilterType] = None,
         index: Optional[str] = None,
         only_documents_without_embedding: bool = False,
         headers: Optional[Dict[str, str]] = None,
+        namespace: Optional[str] = None,
+        type_metadata: Optional[str] = None,
```

Comment:
I'm not sure we can add parameters here, the `BaseDocumentStore` doesn't declare them. 
I mean we technically can, I just don't know whether this will break some other stuff. 🤔

Reply:
The changes in the signature of abstract methods from `BaseDocumentStore` class are done according to the previous implementation of `PineconeDocumentStore` class. Previously, some of abstract methods were changed by adding additional parameters that are necessary for handling different documents inside the document store. 

Here is the full list of changed methods and parameters that were added in previous version of PineconeDocumentStore implementation:

- `write_documents` - _labels_ parameter
- `get_all_documents` - _namespace_ parameter
- `get_all_documents_generator` - _namespace_ parameter
- `get_document_by_id` - _return_embedding_ and _namespace_ parameters
- `get_documents_by_id` - _return_embedding_ and _namespace_ parameters
- `query_by_embedding` - _namespace_ parameter
- `delete_documents` - _drop_ids_ and _namespace_ parameters
- `delete_labels` - _batch_size_ parameter

Accordingly, during this refactoring, some of those methods have been changed by adding additional parameter _type_metadata_ when necessary, in order to support handling different documents using metadata filtering instead of namespace.